packages/@aws-cdk/cloudformation-diff/lib/diff/template-and-changeset-diff-merger.ts

```diff
@@ -0,0 +1,139 @@
+      if (resourceChange.ResourceChange?.LogicalResourceId === undefined) {
+        continue; // Being defensive, here.
+      }
+
+      const propertyReplacementModes: types.PropertyReplacementModeMap = {};
+      for (const propertyChange of resourceChange.ResourceChange.Details ?? []) { // Details is only included if resourceChange.Action === 'Modify'
+        if (propertyChange.Target?.Attribute === 'Properties' && propertyChange.Target.Name) {
+          propertyReplacementModes[propertyChange.Target.Name] = {
+            replacementMode: TemplateAndChangeSetDiffMerger.determineChangeSetReplacementMode(propertyChange),
+          };
+        }
+      }
+
+      changeSetResources[resourceChange.ResourceChange.LogicalResourceId] = {
+        resourceWasReplaced: resourceChange.ResourceChange.Replacement === 'True',
+        resourceType: resourceChange.ResourceChange.ResourceType ?? TemplateAndChangeSetDiffMerger.UNKNOWN_RESOURCE_TYPE, // DescribeChangeSet doesn't promise to have the ResourceType...
+        propertyReplacementModes: propertyReplacementModes,
+      };
+    }
+
+    return changeSetResources;
+  }
+
+  public overrideDiffResourceChangeImpactWithChangeSetChangeImpact(logicalId: string, change: types.ResourceDifference) {
+    // resourceType getter throws an error if resourceTypeChanged
+    if ((change.resourceTypeChanged === true) || change.resourceType?.includes('AWS::Serverless')) {
+      // CFN applies the SAM transform before creating the changeset, so the changeset contains no information about SAM resources
+      return;
+    }
+    change.forEachDifference((type: 'Property' | 'Other', name: string, value: types.Difference<any> | types.PropertyDifference<any>) => {
+      if (type === 'Property') {
+        if (!this.changeSetResources[logicalId]) {
+          (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.NO_CHANGE;
+          (value as types.PropertyDifference<any>).isDifferent = false;
+          return;
+        }
+
+        const changingPropertyCausesResourceReplacement = (this.changeSetResources[logicalId].propertyReplacementModes ?? {})[name]?.replacementMode;
+        switch (changingPropertyCausesResourceReplacement) {
+          case 'Always':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.WILL_REPLACE;
+            break;
+          case 'Never':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.WILL_UPDATE;
+            break;
+          case 'Conditionally':
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.MAY_REPLACE;
+            break;
+          case undefined:
+            (value as types.PropertyDifference<any>).changeImpact = types.ResourceImpact.NO_CHANGE;
+            (value as types.PropertyDifference<any>).isDifferent = false;
+            break;
+          // otherwise, defer to the changeImpact from the template diff
```

Comment:
I just want to refactor in this code change, so this is just a copy of what already exists being lifted into this new file. Not sure what would be an otherwise situation, though

Reply:
Got it. It makes sense to leave it as is for now.